docs: add Role-Based Access Control page; refresh permissions docs (#23133)
mlflow/mlflow: docs/docs/self-hosting/security/role-based-access-control.mdx

Reviewer: Thanks for the detailed write-up! Overall the guide appears to expose a bit too much internal implementation imo. Can we reframe the description to be less technical?

I think [this guide](https://langfuse.com/docs/administration/rbac#role-based-access-controls-in-langfuse) is a great example, it does not expose lots of implementation details but still give necessary info for admins and developers.

For example,
* from ``"Perissions" lives as rows under a role: (resource_type, resource_pattern, permission)``
* to`"Permissions" decides which actions a role can take on a particular resource.`

Author: Fair feedback — the current page reads as "internal architecture with usage examples grafted on" rather than "operator's guide". Will do a structural reframe in the next round to lead with concepts (Roles / Permissions / Workspaces and how they compose) and pull the implementation details (`__user_<id>__` synthetic role, resolver mechanics, `max_permission` semantics) into a single optional deep-dive section near the end. The shipped implementation work this round (mlflow/mlflow#23247, #23248) means the API surface won't keep moving under us, so this is a good time for the reframe.

🤖 *Claude response*

- **Roles** are workspace-scoped. A role named `editor` in workspace `foo` is a
  separate row from `editor` in workspace `bar`.
- **Permissions** live as rows under a role: `(resource_type, resource_pattern,
permission)`. The pattern is either a literal resource id or `*` (apply to all
  resources of that type in the workspace).
- **Direct per-user grants** — when you want to give one user one permission, no
  role to share — are stored under a hidden synthetic role named
  `__user_<id>__`. Same storage, same resolver; the synthetic role is just an
  authoring detail.

### Permission levels

| Permission       | Can read | Can use | Can update | Can delete | Can manage |
| ---------------- | -------- | ------- | ---------- | ---------- | ---------- |
| `READ`           | Yes      | No      | No         | No         | No         |
| `USE`            | Yes      | Yes     | No         | No         | No         |
| `EDIT`           | Yes      | Yes     | Yes        | No         | No         |
| `MANAGE`         | Yes      | Yes     | Yes        | Yes        | Yes        |
| `NO_PERMISSIONS` | No       | No      | No         | No         | No         |

> 

Author: Fixed in [fa2dd3499](https://github.com/mlflow/mlflow/pull/23133/commits/fa2dd3499). Removed the `NO_PERMISSIONS` row from the grantable-permissions table and added a paragraph below explaining that it is the deny-by-default sentinel — not grantable as a role permission or direct grant (the auth server's `_validate_permission_for_resource_type` rejects it). To deny access, simply don't grant the permission rather than try to set `NO_PERMISSIONS` explicitly.

🤖 *Claude response*

```

In the admin UI: `/admin` → Roles → **Create role** → in the _Permissions_
section add `experiment:42 → EDIT`; in the _Assigned users_ section add
`alice` → Create.

### 3. Give a team READ access to every experiment in a workspace

A wildcard pattern lets the role apply to resources that don't exist yet —
adding a new experiment automatically inherits the grant.

```python
role = auth_client.create_role(name="experiment-reader", workspace="ml-research")
auth_client.add_role_permission(
    role_id=role.id,
    resource_type="experiment",
    resource_pattern="*",
    permission="READ",
)

for member in ("alice", "bob", "carol"):
    auth_client.assign_role(username=member, role_id=role.id)
```

In the admin UI: `/admin` → Roles → **Create role** → resource type
`experiment`, pattern `*` (rendered as "All experiments"), permission `READ`;
add the team in the _Assigned users_ section → Create.

### 4. Make a user a workspace admin

Each newly created workspace is automatically seeded with a default `admin`
role (and a `user` role) — see [Default roles](#default-roles). Promoting a
user to workspace admin means assigning the seeded `admin` role:

```python

> admin_role = auth_client.get_role_by_name(workspace="ml-research", name="admin")

Author: Fixed in [fa2dd3499](https://github.com/mlflow/mlflow/pull/23133/commits/fa2dd3499). Same `list_roles` + filter pattern:

```python
admin_role = next(r for r in auth_client.list_roles("ml-research") if r.name == "admin")
auth_client.assign_role(username="alice", role_id=admin_role.id)
```

🤖 *Claude response*

endpoints are gone.

| Removed (legacy)                                                                       | Replacement                                                                                                                                |
| -------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------ |
| `POST /mlflow/experiments/permissions/create`                                          | `POST /mlflow/roles/create` + `POST /mlflow/roles/permissions/add` + `POST /mlflow/roles/assign` (or a direct permission via the admin UI) |
| `POST/GET/PATCH/DELETE` on `/mlflow/{registered-models,scorers,gateway/*}/permissions` | Same role-API pattern                                                                                                                      |
| `POST/GET/PATCH/DELETE` on `/mlflow/workspaces/<workspace>/permissions`                | Assign the seeded `admin` or `user` role for the workspace, or a custom role                                                               |

> | `auth_client.create_experiment_permission()` and the ~30 sibling client methods        | `auth_client.create_role()`, `add_role_permission()`, `assign_role()`, etc.                                                                |

Author: Fixed in [fa2dd3499](https://github.com/mlflow/mlflow/pull/23133/commits/fa2dd3499). The migration table now has three columns — *Pre-RBAC surface*, *Status*, *Replacement* — and explicitly distinguishes **Deprecated** (per-resource permission endpoints + `create_experiment_permission()` family — still work, emit a warning) from **Removed** (workspace-permission endpoints + the four `*_workspace_permission` client methods — calls 404). A short paragraph below the table calls out that the per-resource path is deprecated-not-removed so existing scripts keep working.

🤖 *Claude response*

**Behavioral note: owner delegation.** Pre-RBAC, a resource creator with
`MANAGE` could delegate permissions on that resource to others. Post-RBAC,
delegation requires a workspace admin (any role with `(workspace, *, MANAGE)`)
or a super admin. If your deployment relied on owner-delegation, either
elevate the relevant users to workspace admins or treat permission delegation
as an admin operation going forward.

## API reference

- <APILink fn="mlflow.server.auth.client.AuthServiceClient">`AuthServiceClient` Python API</APILink>
- <Link to="/api_reference/auth/rest-api.html" target="_blank"><span>Authentication REST API</span></Link>

The role-related methods on `AuthServiceClient`:

```python
auth_client.create_role(name, workspace, description=None)
auth_client.get_role(role_id)
auth_client.get_role_by_name(workspace, name)
auth_client.list_roles(workspace)
auth_client.update_role(role_id, name=None, description=None)

> auth_client.delete_role(role_id)

Author: Fixed in [fa2dd3499](https://github.com/mlflow/mlflow/pull/23133/commits/fa2dd3499). The actual signature in `client.py:299` is `create_role(workspace, name, description=None)`. Reordered the API reference accordingly. The example earlier in the page already used keyword arguments, so it stays correct.

🤖 *Claude response*

  role to share — are stored under a hidden synthetic role named
  `__user_<id>__`. Same storage, same resolver; the synthetic role is just an
  authoring detail.

### Permission levels

The grantable permission levels for resource-scoped permissions are:

| Permission | Can read | Can use | Can update | Can delete | Can manage |
| ---------- | -------- | ------- | ---------- | ---------- | ---------- |
| `READ`     | Yes      | No      | No         | No         | No         |
| `USE`      | Yes      | Yes     | No         | No         | No         |
| `EDIT`     | Yes      | Yes     | Yes        | No         | No         |
| `MANAGE`   | Yes      | Yes     | Yes        | Yes        | Yes        |

`USE` is for consuming a resource without modifying it (invoking a gateway
endpoint, referencing a model definition, creating new experiments /
registered models within a workspace).

`NO_PERMISSIONS` exists as the deny-by-default sentinel — it's the effective
state of any `(user, resource)` pair with no matching grant. It is **not
grantable** as a role permission or a direct grant; the auth server rejects
attempts to assign it. To deny access, simply don't grant it; an absent grant

> combined with a non-permissive `default_permission` yields the same effect.

Author: Same fix in [ec7f7b067](https://github.com/mlflow/mlflow/pull/23133/commits/ec7f7b067). Dropped the misleading reference to `default_permission` (which only applies when workspaces are disabled — workspaces are required for RBAC). The paragraph now states explicitly that there is no explicit-deny grant and that a workspace-level grant cannot be excepted by an absent per-resource grant — the resolver finds no specific row, falls through to the workspace grant, and allows. Restricting a single resource means granting narrowly in the first place.

🤖 *Claude response*

import Link from "@docusaurus/Link";
import { APILink } from "@site/src/components/APILink";

# Role-Based Access Control (RBAC)

MLflow's role-based access control lets you express authorization as named, reusable
roles instead of one-off per-resource grants. A **role** is a bundle of permissions;
**users** are assigned to roles; permissions decisions are made against the
combined set of permissions a user holds in the workspace they're acting in.

RBAC is the authorization model for all permission grants. Even a single
one-off grant on one resource goes through it (see [Scenario 1](#1-give-alice-edit-on-experiment-42-one-off)).

> A named role pays off when you have more than a handful of users, when teams

Reviewer: This part sounds like we are imposing something that is not worth ("we mandate role but it only pays off when many users exist"). Can we rephrase it?

Author: Reread this section with your framing — I agree the current wording reads as "you must adopt roles to use auth". Will rephrase to lead with the simple case (one-off direct grants via `grant_user_permission` or the UI's Direct permissions section, no role to author) and frame roles as the natural shape once permission sets are shared. Holding the rewrite for the next round so we can also incorporate your overall framing comment at rbac.mdx:1.

🤖 *Claude response*

share access patterns ("data scientists can read all experiments in
`ml-research`"), or when you want to delegate workspace administration
without giving out global admin. For setting up authentication and managing
users (login, signup, admin status, password updates), see
[Username and Password](/self-hosting/security/basic-http-auth); this page
picks up once users exist.

:::note[Prerequisites]

- Authentication enabled (`mlflow server --app-name basic-auth`)

> - Workspaces enabled (`--enable-workspaces`); RBAC roles are workspace-scoped

Reviewer 2: Doesn't RBAC work when workspace is not enabled?

Author: Good catch — softened the prerequisite in `ae67927d4`. RBAC works against the reserved `default` workspace out of the box; `--enable-workspaces` is only required for multi-workspace setups (named workspaces, the workspace switcher, `/admin/ws?workspace=…`). Verified at `sqlalchemy_store.py:90-102` — when the env var is unset, `_get_active_workspace_name` returns `DEFAULT_WORKSPACE_NAME`.

🤖 *Claude response*

  Pass `--enable-workspaces` to enable multi-workspace setups: named
  workspaces, the workspace switcher, and `/admin/ws?workspace=…` for
  Workspace Managers.

:::

## The model

```mermaid

> flowchart LR

Reviewer: I think it is cleaner to illustrate entity (e.g. users) and relation (user_role_assignments) separately. Also can we include "resources" in the picture?

Author: Agreed — the current diagram conflates user→role assignment with role→permission. Will redraw with users, roles, and resources as three entity boxes plus `user_role_assignments` and `role_permissions` as labeled relations. Holding for next round so we get all three of your diagram-related asks in one pass (this one, rbac.mdx:133, and possibly the rbac.mdx:1 reframe). 🤖 *Claude response*

    Users[users] --> URA[user_role_assignments]
    URA --> Roles[roles]
    Roles --> RP[role_permissions]
```

A user-role assignment links one user to one role; a role contains zero or
more `role_permissions` rows of the form
`(resource_type, resource_pattern, permission)`.

- **Users** are global (one row per identity in the auth database).
- **Roles** are workspace-scoped. A role named `editor` in workspace `foo` is a
  separate row from `editor` in workspace `bar`.
- **Permissions** live as rows under a role: `(resource_type, resource_pattern,
permission)`. The pattern is either a literal resource id or `*` (apply to all
  resources of that type in the workspace).

> - **Direct per-user grants** (when you want to give one user one permission,

Reviewer: It's a bit confusing that the concepts described here does not map to the diagram and its naming 1 to 1. Can we align them and potentially move "Direct per-user grants" to a separate note or a section?

Author: Good catch — synthetic per-user roles are an implementation detail that doesn't belong in the entity-naming flow. Will pull this out into its own subsection (probably a callout titled "Direct grants without a role") and keep the top-level concept list to the three named entities (Roles, Permissions, Workspaces). Holding for the next round, paired with the rbac.mdx:34 diagram redraw. 🤖 *Claude response*

| `EDIT`     | Yes      | Yes     | Yes        | No         | No         |
| `MANAGE`   | Yes      | Yes     | Yes        | Yes        | Yes        |

`USE` is for consuming a resource without modifying it (invoking a gateway
endpoint, referencing a model definition, creating new experiments /
registered models within a workspace).

`NO_PERMISSIONS` exists as the deny-by-default sentinel. It's the effective
state of any `(user, resource)` pair with no matching grant in the resolver's
chain. It is **not grantable** as a role permission or a direct grant; the
auth server rejects attempts to assign it.

Because grants are folded via a max (see
[Permission resolution](#permission-resolution)), RBAC has no
explicit-deny override: an absent or lesser per-resource grant cannot
revoke a broader grant the user already holds. To restrict access to a
specific resource, grant access more narrowly (per-resource or
resource-type wildcard) rather than holding a broad workspace-level grant
and trying to except the resource.

### Workspace-level grants

Workspace-wide access is expressed as a permission on the special
`(resource_type='workspace', resource_pattern='*')` slot. Two levels:

> - `USE` on `(workspace, *)` is the **workspace member** grant. Holders get

Reviewer 2: workspace member -> Workspace Member

Author: Done in `ae67927d4` — capitalized "Workspace Member" everywhere it refers to the UI-aligned identity tier (matching "Workspace Manager"). Updated both `rbac.mdx` and `permissions.mdx`, plus the Default-roles tables on both pages.

🤖 *Claude response*

   no role-derived grant matches. With workspaces disabled, it always applies.
   With workspaces enabled, it applies only to the reserved `default`
   workspace when `grant_default_workspace_access` is set; otherwise the
   effective default is deny.

### Identity tiers

The admin UI uses **Platform Admin** for the first tier and **Workspace
Manager** for the second; the doc uses the same labels.

| Tier              | How it's expressed                          | Capability                                                                     |
| ----------------- | ------------------------------------------- | ------------------------------------------------------------------------------ |
| Platform Admin    | `is_admin = true` on the user row           | Unrestricted system-wide. Sole bearer of user delete and bulk operations.      |
| Workspace Manager | Holds `(workspace, *, MANAGE)` via any role | Full authority within those workspaces; can manage roles, users, and grants.   |
| Regular user      | Any other authenticated identity            | No admin UI access; authorization flows through role-derived permissions only. |

> ## Common scenarios

Reviewer 2: I'd separate the use case and method (UI, API) and explain both the UI and API paths for each primary use case.

Author: Restructured in `ae67927d4`. Each of Scenarios 1-5 now has explicit `**Via the admin UI:**` and `**Via the role API:**` subsections so the reader can scan by intent.

🤖 *Claude response*

Author: Reworked Scenario 1 in 7af69cdb6 along these lines: each scenario now has a "Via the convenience API", "Via the admin UI", and "Via the role API" subsection where applicable, so the use case and the method are clearly separated. The convenience API path is new (`grant_user_permission`) and is now the simplest single-call path for one-off direct grants — it didn't exist when you first reviewed.

🤖 *Claude response*

| ----------------- | ------------------------------------------- | ------------------------------------------------------------------------------ |
| Platform Admin    | `is_admin = true` on the user row           | Unrestricted system-wide. Sole bearer of user delete and bulk operations.      |
| Workspace Manager | Holds `(workspace, *, MANAGE)` via any role | Full authority within those workspaces; can manage roles, users, and grants.   |
| Regular user      | Any other authenticated identity            | No admin UI access; authorization flows through role-derived permissions only. |

> ## Common scenarios

Reviewer: Can we let claude or gemini draw some diagram for comparing three scenario?

Author: Will tackle in the next round. Three scenarios → three small visual flows (one-off direct grant, shared role, workspace floor). Pairs with the rbac.mdx:34 entity-relation diagram redraw. 🤖 *Claude response*

Programmatically, the synthetic role isn't exposed as a separate API surface.
Use the same role API as the reusable case below ([Scenario 2](#2-give-alice-edit-on-experiment-42-reusable)),
just with a single-user role: `create_role` + `add_role_permission` +
`assign_role`. The pre-RBAC `create_experiment_permission()` /
`update_experiment_permission()` family still works for now (it
internally writes to the synthetic role) but emits a deprecation warning
on every call; new code should not rely on it.

### 2. Give Alice EDIT on experiment 42 (reusable)

When the same permission set will be granted to other users (now or later),
build it as a role:

```python
# 1. Create the role (workspace-scoped)
role = auth_client.create_role(name="exp-42-editor", workspace="ml-research")

# 2. Add the permission
auth_client.add_role_permission(
    role_id=role.id,
    resource_type="experiment",
    resource_pattern="42",
    permission="EDIT",
)

# 3. Assign Alice
auth_client.assign_role(username="alice", role_id=role.id)
```

In the admin UI: `/admin` → Roles → **Create role** → in the _Permissions_
section add `experiment:42 → EDIT`; in the _Assigned users_ section add
`alice` → Create.

> ### 3. Give a team READ access to every experiment in a workspace

Reviewer 2: Do we need this example?

Author: Kept it but tightened in `ae67927d4`. The wildcard pattern is a distinct primitive worth showing once (the role applies to resources that don't exist yet), and the multi-user loop is the only place this PR demonstrates that. Trimmed the boilerplate that overlapped with Scenario 2 and made the framing explicit ("the shape is identical to Scenario 2 with two differences: `resource_pattern=\"*\"` and a loop over the team").

🤖 *Claude response*

Author: Fair — let me re-read the section. If you can confirm which example you mean (the Scenario 2 single-user role pattern?), I'll either remove it or restructure. Asking because the line moved a bit after the earlier rework. 🤖 *Claude response*

Author: The restructure in `9617f3b09` makes the wildcard scenario the canonical team-access pattern (Scenario 2). It carries the per-scenario mermaid + Tabs and demonstrates `resource_pattern="*"` so readers see the wildcard primitive once. Happy to drop it if you still think it doesn't earn its place — wanted to take one more pass first. 🤖 *Claude response*

### 5. Onboard a team of N users with the same access

Either reuse the seeded `<workspace>/user` role or create one once. Two entry
points to add multiple users:

- **From the role.** `/admin` → Roles → click the role → **Edit role** → in the
  _Assigned users_ section, add all the users → review → apply.
- **From each user.** Users → click user → **Edit access** → in the _Role
  assignments_ section, add the role → apply.

The first form is one round trip per user assignment; the second is one round
trip per user end-to-end. Pick whichever fits the operator's mental model. The
result is identical.

## Admin UI

The admin UI is the operator-facing surface for everything above. It's reached
from two entry points:

- **Platform Admins** (`is_admin = true`) navigate to `/admin` via the
  sidebar `Manage` entry. The page renders cross-workspace: every user in
  the system, every role in every workspace.
- **Workspace Managers** click a gear icon on the home-page workspaces table
  next to a workspace they administer. The link lands at
  `/admin/ws?workspace=<name>`, the per-workspace view scoped to roles and
  users they can see.

> Both views share the same layout: a **Users** tab and a **Roles** tab.

Reviewer 2: Can we add a screenshot?

Reviewer: +1

Author: Capturing this in a follow-up. Will post here once the screenshot is in. 🤖 *Claude response*

Author: Screenshots landed in `b491c3de2` and are preserved in the latest restructure (`9617f3b09`) — `Managing Users` and `Managing Roles` each have their picture(s) now. Resolving. 🤖 *Claude response*

Author: Screenshots landed in `b491c3de2` and are preserved through the restructure (`9617f3b09`). 🤖 *Claude response*

### Users tab

Lists every user, each row showing their visible roles. Click a username to
open the user detail page, then **Edit access** to manage that user's roles,
direct permissions, and admin status (Platform Admins only). Changes are

> previewed in a Review step before they're applied.

Reviewer 2: Can we add a screenshot?

Author: Capturing this in a follow-up. Will post here once the screenshot is in. 🤖 *Claude response*

Author: Screenshots landed in `b491c3de2` and are preserved in the latest restructure (`9617f3b09`) — `Managing Users` and `Managing Roles` each have their picture(s) now. Resolving. 🤖 *Claude response*

Author: Screenshots landed in `b491c3de2` and are preserved through the restructure (`9617f3b09`). 🤖 *Claude response*

### Roles tab

Lists roles in the active scope (all roles for Platform Admins; the active
workspace's roles for Workspace Managers). **Create role** opens a
single-page form with three sections (_Role details_, _Permissions_,
_Assigned users_) and submits them in one shot. **Edit role** on the role

> detail page mirrors the same shape.

Reviewer 2: Can we add a screenshot?

Author: Capturing this in a follow-up. Will post here once the screenshot is in. 🤖 *Claude response*

Author: Screenshots landed in `b491c3de2` and are preserved in the latest restructure (`9617f3b09`) — `Managing Users` and `Managing Roles` each have their picture(s) now. Resolving. 🤖 *Claude response*

Author: Screenshots landed in `b491c3de2` and are preserved through the restructure (`9617f3b09`). 🤖 *Claude response*

| `POST/GET/PATCH/DELETE` on `/mlflow/workspaces/<workspace>/permissions`                                                                        | **Removed**                                                                                         | Assign the seeded `admin` or `user` role for the workspace, or a custom role with a `(workspace, *, …)` permission                        |
| `auth_client.set_workspace_permission()`, `list_workspace_permissions()`, `delete_workspace_permission()`, `list_user_workspace_permissions()` | **Removed**                                                                                         | The role API (`assign_role` to a workspace-scoped role with a `(workspace, *, …)` permission)                                             |

The per-resource path is "deprecated, not removed" so existing scripts keep
working; new code should use the role API directly. The workspace-permission
path is fully gone; calls to those endpoints return 404.

> **Behavioral note: owner delegation.** Pre-RBAC, a resource creator with

Reviewer 2: Sorry for missing this change during reviews, but why did we make this change? I think users with "MANAGE" access should continue to be able to control the permission for the specific resource

Author: Hmm, true. I think the main reason is that we lack the UI to do that (We only see manage UI for certain admins.) This actually makes MANAGE kinda useless until we have a proper UI/API for it.

While we don't have UI for it yet, what do you think of providing an API for users with MANAGE permissions? (So that's back-end changes only.)

That would better for backward-compatibility. WDYT?

Author: Follow-up to my earlier reply: rather than gate this on a future API, I went and checked the code (`mlflow/server/auth/__init__.py:1979-2021`). The deprecated per-resource endpoints are still gated by per-resource MANAGE — so during the deprecation window per-resource MANAGE holders **can** still delegate via `/mlflow/experiments/permissions/*` etc. The new role API (`add_role_permission`, `assign_role`) is the only path that requires workspace MANAGE / Platform Admin.

Dropped the misleading "Behavioral note: owner delegation" paragraph in `ae67927d4` and replaced it with one that's accurate about both the current state (per-resource MANAGE still delegates via the deprecated endpoints) and the future state (when those are removed, only workspace MANAGE will delegate). Also added a test to pin the deprecated-vs-role-API gating in #23219.

🤖 *Claude response*

Reviewer 2: >what do you think of providing an API for users with MANAGE permissions?

If users with MANAGE permissions can give/remove permissions to the specific resources through API, I think that's sufficient for now.

Author: Good catch — the answer is yes, they still can, just via a different surface than the legacy `create_*_permission()` family. The shipped convenience API ([mlflow/mlflow#23247](https://github.com/mlflow/mlflow/pull/23247)) preserves this **permanently**:

```python
auth_client.grant_user_permission(username, resource_type, resource_id, permission)
```

`grant_user_permission` is gated by `validate_can_manage_resource`, which checks per-resource MANAGE on the target — exactly the same authority check the legacy methods used. So a user with `(experiment, 42, MANAGE)` can still grant other users access to experiment 42, without holding workspace-wide MANAGE.

The docs migration table and Scenario 1 (8452fad8e, 2894bad4d) both call this out explicitly. The role API itself (`create_role`, `add_role_permission`, `assign_role`) stays admin-only — that part is unchanged from the legacy world. 🤖 *Claude response*

`MANAGE` could delegate permissions on that resource to others. Post-RBAC,
delegation requires a Workspace Manager (any role with
`(workspace, *, MANAGE)`) or a Platform Admin. If your deployment relied on
owner-delegation, either elevate the relevant users to Workspace Managers
or treat permission delegation as an admin operation going forward.

## API reference

- <APILink fn="mlflow.server.auth.client.AuthServiceClient">`AuthServiceClient` Python API</APILink>
- <Link to="/api_reference/auth/rest-api.html" target="_blank"><span>Authentication REST API</span></Link>

The role-related methods on `AuthServiceClient`:

> ```python

Reviewer 2: Do we need this section?

Author: Trimmed in `ae67927d4`: dropped the manually-maintained method-signature code block. Kept only the two `<APILink>` / `<Link>` references that point to the auto-generated API docs.

🤖 *Claude response*

| `POST/GET/PATCH/DELETE` on `/mlflow/workspaces/<workspace>/permissions`                                                                        | **Removed** | Assign the seeded `admin` or `user` role for the workspace, or a custom role with a `(workspace, *, …)` permission                                                               |
| `auth_client.set_workspace_permission()`, `list_workspace_permissions()`, `delete_workspace_permission()`, `list_user_workspace_permissions()` | **Removed** | The role API (`assign_role` to a workspace-scoped role with a `(workspace, *, …)` permission)                                                                                    |

All four legacy surfaces are removed at the same upgrade that lands RBAC.
There is no deprecation-warning window: basic-auth was still marked
experimental, so the carry cost of 25 deprecation-warning-emitting
methods outweighed the soft-transition benefit. The auth-store backfill
migration moves existing legacy permission rows into synthetic-role
storage, so existing grants survive the upgrade — only the API surface

> changes.

Author: The "Removed" status describes the state after #23247 (adds the `grant_user_permission` / `revoke_user_permission` / `check_user_permission` convenience APIs) and #23337 (removes the deprecated legacy per-resource methods + endpoints) land. Both target master and will merge before this docs PR ships, so the table will be accurate by the time readers see it. 🤖 *Claude response*

import Link from "@docusaurus/Link";
import { APILink } from "@site/src/components/APILink";

# Role-Based Access Control (RBAC)

MLflow's role-based access control is the authorization layer for every
permission grant — but the operator surface scales with what you need:

> - **Give one user one permission?** A single

Reviewer: Can we explain "Create a role" pass as a single primary path, and just keep `grant_user_permission` as a tip box? The former is the core of RBAC idea, and having two other options here upfront makes it less clear.

Author: Done in `9617f3b09`. Intro is now a single paragraph framing roles as the model; the one-off case lives in a `:::tip` pointing at the `### Direct grants without a role` subsection (which also absorbed the operational quick-reference from old Scenario 1). 🤖 *Claude response*

  [`grant_user_permission`](#1-give-alice-edit-on-experiment-42-one-off)
  call (or one click in the admin UI). No role to author, no users to
  assign.
- **Share the same permission set across users?**
  [Create a role](#2-give-alice-edit-on-experiment-42-reusable), add
  the permissions, assign the users. The role lives in one place;
  changes propagate automatically.
- **Make someone a workspace admin?**
  [Assign the seeded `admin` role](#4-make-a-user-a-workspace-manager)
  for that workspace.

Roles pay off when patterns repeat or you need to delegate workspace
administration. For one-off grants, the convenience APIs cover the
simple case directly. Both flows write to the same underlying model, so
you can mix and match as access patterns evolve.

For setting up authentication and managing users (login, signup, admin
status, password updates), see
[Username and Password](/self-hosting/security/basic-http-auth); this
page picks up once users exist.

:::note[Prerequisites]

- Authentication enabled (`mlflow server --app-name basic-auth`)

> - RBAC works against the reserved `default` workspace out of the box.

Reviewer: Is this prerequisite? Can we move down to workspace section?

Author: Done in `9617f3b09`. The workspace bullet is out of the prerequisite block; multi-workspace setup is now covered in the new `## Workspace isolation` section. 🤖 *Claude response*

  Pass `--enable-workspaces` to enable multi-workspace setups: named
  workspaces, the workspace switcher, and `/admin/ws?workspace=…` for
  Workspace Managers.

:::

## The model

Three entities, two relations:

```mermaid
flowchart LR
    Users[Users]
    Roles[Roles]
    Resources[Resources]

    Users -- user_role_assignments --> Roles
    Roles -- role_permissions --> Resources
```

- **Users** are global (one row per identity in the auth database).

> - **Roles** are workspace-scoped bundles of permissions. A role named

Reviewer: Would it be clearer if we just have `Workspace Isolation` and add one sentence there everything is bounded by workspace if enabled? Each section is like 50% longer than minimum to include workspace handling😅

It might be confusing for users who haven't opted in to Workspace experience. Extracting the diff between workspace on and off to a single place would probably simpler and more readable.

Author: Done in `9617f3b09`. New top-level `## Workspace isolation` section consolidates `--enable-workspaces` behavior, default workspace fallback, and the `(workspace, *)` USE/MANAGE grants. Scenarios no longer re-explain workspace handling — they just note "multi-workspace mode" once in the lead-in. 🤖 *Claude response*

   no role-derived grant matches. With workspaces disabled, it always applies.
   With workspaces enabled, it applies only to the reserved `default`
   workspace when `grant_default_workspace_access` is set; otherwise the
   effective default is deny.

### Identity tiers

The admin UI uses **Platform Admin** for the first tier and **Workspace
Manager** for the second; the doc uses the same labels.

| Tier              | How it's expressed                          | Capability                                                                     |
| ----------------- | ------------------------------------------- | ------------------------------------------------------------------------------ |
| Platform Admin    | `is_admin = true` on the user row           | Unrestricted system-wide. Sole bearer of user delete and bulk operations.      |
| Workspace Manager | Holds `(workspace, *, MANAGE)` via any role | Full authority within those workspaces; can manage roles, users, and grants.   |
| Regular user      | Any other authenticated identity            | No admin UI access; authorization flows through role-derived permissions only. |

> ## Common scenarios

Reviewer: How each of these model map to the 5 scenarios? Can we have one illustration per scenario (and embed within the sub-section)?

Author: Done in `9617f3b09`. The big 3-shape diagram is gone; each scenario now has its own small mermaid embedded in the subsection. 🤖 *Claude response*

        B2[Bob] --> B3
        B3 -->|role_permissions| B4[experiment 42<br/>EDIT]
    end

    subgraph floor[Workspace floor — many users, workspace-wide]
        direction LR
        C1[Alice] --> C3[ml-research/user<br/>seeded role]
        C2[Bob] --> C3
        C3 -->|role_permissions| C4[workspace *<br/>USE]
    end
```

The direct-grant case uses the per-user `__user_<id>__` synthetic role
under the hood (Alice's row never names a shared role); the shared and
workspace-floor cases use an operator-authored or seeded role
respectively. The resolver folds whichever grants apply via `max` —
nothing changes about how permissions are evaluated.

The examples below assume you've authenticated as a Platform Admin or as a
Workspace Manager of `ml-research`.

```python
from mlflow.server import get_app_client

tracking_uri = "http://localhost:5000"
auth_client = get_app_client("basic-auth", tracking_uri=tracking_uri)
```

### 1. Give Alice EDIT on experiment 42 (one-off)

For a one-user, one-resource grant that won't be reused, a **direct
permission** is the simplest fit. Internally it lives on Alice's synthetic
per-user role; no admin-managed role to maintain.

> **Via the admin UI:** `/admin` → Users → click `alice` → **Edit access** → in

Reviewer: Let's use tabs for UI/API split, and replace "... -> ... -> " with a numbered bullet list for better readability.

Author: Done in `9617f3b09`. Each scenario uses `<Tabs groupId="rbac-flow">` to split **Admin UI** / **Role API**; the UI flows are numbered lists. Scenario 4 (onboarding N users) has three tabs because there are two equally-valid UI flows. 🤖 *Claude response*

  the _Assigned users_ section, add all the users → review → apply.
- **From each user.** Users → click user → **Edit access** → in the _Role
  assignments_ section, add the role → apply.

The first form is one round trip per user assignment; the second is one
round trip per user end-to-end. Pick whichever fits the operator's mental
model; the result is identical.

**Via the role API:** loop over the team and call `assign_role` once per
member, as in Scenario 3.

## Admin UI

The admin UI is the operator-facing surface for everything above. It's reached
from two entry points:

- **Platform Admins** (`is_admin = true`) navigate to `/admin` via the
  sidebar `Manage` entry. The page renders cross-workspace: every user in
  the system, every role in every workspace.
- **Workspace Managers** click a gear icon on the home-page workspaces table
  next to a workspace they administer. The link lands at
  `/admin/ws?workspace=<name>`, the per-workspace view scoped to roles and
  users they can see.

Both views share the same layout: a **Users** tab and a **Roles** tab.

![Admin UI showing Users and Roles tabs](/images/self-hosting/rbac-admin-ui-overview.png)

> ### Users tab

Reviewer: The order of screenshot and the section is a bit confusing: `[Users tab picture] -> ### Users tab -> [User detail page picture] -> ### Roles tab -> [Roles tab picture]`. Can we move the first picture down to `Users tab` section?

Author: Done in `46121a289`. The overview screenshot now sits under `### Managing Users` so each picture follows its section heading. 🤖 *Claude response*

Lists every user, each row showing their visible roles. Click a username to

> open the user detail page, then **Edit access** to manage that user's roles,

Reviewer: ```suggestion
### Managing Users

Click a username in the `Users` tab to manage roles assigned to each user.
```

I think it is cleaner to name by the purpose not the page name.

Author: Renamed to `### Managing Users` in `46121a289` (kept the existing lead-in about Edit access / Review step since that detail still seemed worth surfacing). 🤖 *Claude response*

direct permissions, and admin status (Platform Admins only). Changes are
previewed in a Review step before they're applied.

![User detail page showing role assignments](/images/self-hosting/rbac-users-tab-roles.png)

> ### Roles tab

Reviewer: ```suggestion
### Managing Roles
```

Author: Renamed to `### Managing Roles` in `46121a289`. 🤖 *Claude response*

Lists roles in the active scope (all roles for Platform Admins; the active
workspace's roles for Workspace Managers). **Create role** opens a
single-page form with three sections (_Role details_, _Permissions_,
_Assigned users_) and submits them in one shot. **Edit role** on the role
detail page mirrors the same shape.

![Roles tab listing roles across workspaces](/images/self-hosting/rbac-roles-tab.png)

![Edit role form with Role details and Permissions sections](/images/self-hosting/rbac-edit-role-form.png)

## Default roles

When `MLFLOW_RBAC_SEED_DEFAULT_ROLES` is on (it is, by default), MLflow seeds
two roles into every newly created workspace:

| Role    | Permission               | Intent                                                                                                 |
| ------- | ------------------------ | ------------------------------------------------------------------------------------------------------ |

> | `admin` | `(workspace, *, MANAGE)` | Workspace Manager. Full authority within the workspace.                                                |

Reviewer: This feels a bit confusing... above we use `Platform Admin` and `Workspace Manager`, but here we say `admin = Workspace Manager`. Maybe we should just say "Workspace Admin"?

Also isolation workspace delta into a section as suggested above should help - `admin` is full authority, `user` is read+create, then everything is gated by workspace if enabled.

Author: Renamed to **Workspace Admin** throughout in `9617f3b09` (Identity tiers, Default roles "Intent" column, Admin UI entry points, scenario heading). The workspace on/off delta is extracted into `## Workspace isolation` per your other comment. 🤖 *Claude response*

| `user`  | `(workspace, *, USE)`    | Workspace Member. Reads every resource in the workspace; can create experiments and registered models. |

The user who creates the workspace is automatically assigned the seeded `admin`
role for that workspace.

To disable the seeding (for installations that prefer to define roles
manually):

```bash
export MLFLOW_RBAC_SEED_DEFAULT_ROLES=false
```

> ## Migrating from the legacy permission model

Reviewer: Can we add a small section for rolling back with the downgrade command?

Author: Added in `9617f3b09` as `### Rollback`: covers the `mlflow db downgrade` command, the one-release-cycle window for legacy tables on disk, and the re-apply behavior on re-upgrade. 🤖 *Claude response*

per-resource permission endpoints, the auth-store backfill migration
translates the legacy tables
(`experiment_permissions`, `registered_model_permissions`, the four
`gateway_*_permissions`, `scorer_permissions`, `workspace_permissions`) into
`role_permissions` rows. The legacy tables remain on disk for rollback safety
through at least one full release cycle, then are dropped in a subsequent
migration.

The wire surface and `AuthServiceClient` shape change as follows:

| Pre-RBAC surface                                                                                                                               | Status      | Replacement                                                                                                                                                                      |
| ---------------------------------------------------------------------------------------------------------------------------------------------- | ----------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `POST/GET/PATCH/DELETE` on `/mlflow/{experiments,registered-models,scorers,gateway/*}/permissions`                                             | **Removed** | `POST /mlflow/users/permissions/grant` / `revoke` for one-user one-resource direct grants, or the role API (`POST /mlflow/roles/create` + `add` + `assign`) for shared sets.     |

Reviewer: This table includes too much low-level code imo, can we simplify it? For example, can we say `workspace-level permission APIs such as set_workspace_permission`, instead of listing all workspace APIs?

Author: Done in `9617f3b09`. The table is now two category rows ("Per-resource permission REST endpoints + `AuthServiceClient` methods" / "Workspace permission REST endpoints + `AuthServiceClient` methods") instead of the four-row enumeration. 🤖 *Claude response*

import Link from "@docusaurus/Link";
import { APILink } from "@site/src/components/APILink";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Role-Based Access Control (RBAC)

MLflow's role-based access control expresses permissions through **roles** —
named bundles of `(resource_type, resource_pattern, permission)` grants you
assign to users. A role lives in one place; changes propagate to every user
assigned to it. When workspaces are enabled (`--enable-workspaces`), roles

> and their grants are scoped per workspace.

Reviewer: ```suggestion
```suggestion
In MLflow, you give users permissions by assigning them **roles**. A role is a named group of permissions - you define it once, assign it to anyone who needs the same access, and update it in one place when things change. If you turn on workspaces (`--enable-workspaces`), each role only applies inside the workspace it was created in.
```
```

:::tip[One-off grants]

For a single-user single-resource grant — no shared pattern, nothing
reusable — skip authoring a role and use the
[`grant_user_permission`](#direct-grants-without-a-role) convenience API
(or the **Direct permissions** section of the admin UI). Internally it
writes to the same role-based model.

:::

For setting up authentication and managing users (login, signup, admin
status, password updates), see
[Username and Password](/self-hosting/security/basic-http-auth); this page
picks up once users exist.

:::note[Prerequisite]

Authentication enabled (`mlflow server --app-name basic-auth`).

:::

> ## The model

Reviewer: ```suggestion
## Concepts
```

nit: "model" is very overloaded in MLflow context

Three entities, two relations. Roles (and the grants they carry) are
workspace-scoped when workspaces are enabled; see
[Workspace isolation](#workspace-isolation) for the on/off delta.

```mermaid
flowchart LR
    Users[Users]
    Roles[Roles]
    Resources[Resources]

    Users -- user_role_assignments --> Roles
    Roles -- role_permissions --> Resources
```

- **Users** are global (one row per identity in the auth database).
- **Roles** are workspace-scoped bundles of permissions. A role named
  `editor` in workspace `foo` is a separate row from `editor` in workspace
  `bar`. See [Workspace isolation](#workspace-isolation) for how scoping
  behaves with and without `--enable-workspaces`.
- **Resources** are experiments, registered models, prompts, scorers, and
  AI Gateway resources (secrets, endpoints, model definitions). A role
  grants permissions on a resource via a `(resource_type, resource_pattern,
permission)` row, where the pattern is either a literal resource id or `*`

>   (apply to all resources of that type in the workspace).

Reviewer: ```suggestion
- **User** — an actor who wants to access resources.
- **Resource** — a target entity in MLflow that permissions are managed on, such as experiments, registered models, prompts, scorers, or AI Gateway resources (secrets, endpoints, model definitions).
- **Role** — a named bundle of permissions on resources, for example `EDIT` access to all experiments. Users gain access to a resource by being assigned a role that grants the needed permission on it.
```

Let's explain the core concept rather than workspace boundary. The workspace session is already linked above.

> ### Direct grants without a role

Reviewer: Can we move this down, maybe after default role? It is confusing that the page is RBAC but we start from a situation where users bypass role setup.

For the one-user one-resource case, you don't need to author a single-user
role. Each user has a reserved per-user role that holds their direct
grants; the `grant_user_permission` / `revoke_user_permission` APIs and the
admin UI's _Direct permissions_ section write to it directly. From an
operator's perspective it's one call or one form field — the reserved role
is an implementation detail you never touch.

**Admin UI:** `/admin` → **Users** → click the user → **Edit access** → in
the _Direct permissions_ section, add the `(resource, permission)` grant →
review → apply.

**API:**

```python
auth_client.grant_user_permission("alice", "experiment", "42", "EDIT")
```

`grant_user_permission` is gated by per-resource MANAGE on the target
resource — the same authority check the legacy
`create_experiment_permission()` family used — so an experiment owner who
holds `(experiment, 42, MANAGE)` can grant access without holding
workspace-wide MANAGE. The role API itself requires workspace MANAGE or
Platform Admin, since authoring a role is an administrative action.

> 
> ### Permission levels

Reviewer: ```suggestion
## Role

### Permission levels
```
It is a bit weird that there is no section named "Role" in RBAC page:p

Can we also move "Default roles" section under `## Role`?

explicit-deny override: an absent or lesser per-resource grant cannot
revoke a broader grant the user already holds. To restrict access to a
specific resource, grant access more narrowly (per-resource or
resource-type wildcard) rather than holding a broad workspace-level grant
and trying to except the resource.

### Permission resolution

For each authorization check, MLflow evaluates the user's effective
permission as follows:

1. **Platform Admin bypass**: `is_admin = true` short-circuits to allow.
2. **Role-derived grants**: all of the user's roles in the request's workspace
   contribute. Every grant that applies to the resource is folded together
   via a max, and the highest-priority permission wins
   (`MANAGE` > `EDIT` > `USE` > `READ`). A `(workspace, *, …)` row
   participates in the same max because it applies to all resources in the
   workspace, so a workspace-level grant acts as a floor that per-resource
   grants can upgrade but never downgrade.
3. **Server `default_permission`**: a configured fallback consulted only
   when no role-derived grant matches. See
   [Workspace isolation](#workspace-isolation) for how this interacts with
   the workspace setting.

> ### Identity tiers

Reviewer: ```suggestion
### User tiers
```

nit: Is "user" and "identify" exchangeable here? If so, let's not introduce a new term.

The admin UI uses **Platform Admin** for the first tier and **Workspace
Admin** for the second; the doc uses the same labels.

| Tier            | How it's expressed                          | Capability                                                                     |
| --------------- | ------------------------------------------- | ------------------------------------------------------------------------------ |
| Platform Admin  | `is_admin = true` on the user row           | Unrestricted system-wide. Sole bearer of user delete and bulk operations.      |
| Workspace Admin | Holds `(workspace, *, MANAGE)` via any role | Full authority within those workspaces; can manage roles, users, and grants.   |
| Regular user    | Any other authenticated identity            | No admin UI access; authorization flows through role-derived permissions only. |

> ## Workspace isolation

Reviewer: Can we move this down after "Admin UI"?

Roles are workspace-scoped. The same role name in two workspaces is two
separate rows; permissions evaluated in one workspace don't bleed into
another. How that scoping behaves depends on whether multi-workspace
support is enabled.

> 

Reviewer: ```suggestion
A [workspace](https://mlflow.org/docs/latest/self-hosting/workspaces/) is an
isolated container for MLflow resources. Workspaces are off by default. To
turn them on, start the server with `--enable-workspaces`. Once enabled,
every experiment, registered model, prompt, scorer, and AI Gateway resource
belongs to exactly one workspace, and nothing crosses the boundary.

Roles and permissions follow the same boundary. A role named `editor` in
workspace `foo` is a completely separate row from `editor` in workspace
`bar`, and a permission grant made in one workspace has no effect in any
other. How that boundary shows up depends on whether multi-workspace
support is enabled.
```

Since Workspace is a new feature, let's add a small introduction about it.

first-class. The home page renders a workspace switcher; the admin UI gets
a per-workspace entry point (`/admin/ws?workspace=<name>`); roles must be
created in a specific workspace. The server's `default_permission` falls
back only to the reserved `default` workspace when
`grant_default_workspace_access` is set; otherwise the effective default in
each named workspace is deny.

Two grants on the special `(resource_type='workspace',
resource_pattern='*')` slot define workspace-wide access:

- `USE` on `(workspace, *)` — the **Workspace Member** grant. USE-level
  access (read + use) to every resource in the workspace, plus the ability
  to create experiments and registered models. Per-resource grants can
  upgrade individual rows. A user without this grant gets no workspace
  floor — only the specific resource grants they hold.
- `MANAGE` on `(workspace, *)` — the **Workspace Admin** grant. Full
  authority within the workspace, including creating roles, granting
  permissions, and managing role assignments. Cannot perform system-wide
  operations such as deleting users.

The seeded `admin` and `user` roles each carry one of these grants; see
[Default roles](#default-roles).

> ## Common scenarios

Reviewer: This section is very clear, thanks a lot!

)
for member in ("alice", "bob", "carol"):
    auth_client.assign_role(username=member, role_id=role.id)
```

</TabItem>
</Tabs>

### 3. Make a user a Workspace Admin

Every newly created workspace is seeded with a default `admin` role (and a
`user` role; see [Default roles](#default-roles)). Promoting a user means
assigning them the seeded `admin` role for that workspace. The role
explicitly carries `(workspace, *, MANAGE)`; unlike the legacy workspace
`MANAGE` permission, it does not implicitly fan out to every resource type.

```mermaid
flowchart LR
    Alice[Alice] --> R["ml-research/admin<br/>(seeded)"]
    R --> P["workspace *<br/>MANAGE"]
```

<Tabs groupId="rbac-flow">
<TabItem value="ui" label="Admin UI">

1. Navigate to `/admin` → **Users**.
2. Click `alice`.
3. Click **Edit access**.
4. In _Role assignments_, add `ml-research/admin`.
5. Review changes → apply.

</TabItem>
<TabItem value="api" label="Role API">

```python
# The seeded ``admin`` role lives in the workspace; look it up by name.
admin_role = next(r for r in auth_client.list_roles("ml-research") if r.name == "admin")
auth_client.assign_role(username="alice", role_id=admin_role.id)
```

</TabItem>
</Tabs>

> ### 4. Onboard a team of N users with the same access

Reviewer: Is this recommended option? I feel like this should be done by a role (like scenario #2) rather than user-role?

Author: I think it's fair to alter `user` role to the workspace creator's needed. But I do agree that the recommend option should be another role. Let me fix some wordings.

Grants can target a single resource (`experiment:42`) or every resource of
a type (`experiment:*`), so one role can express both fine-grained and broad
access. For example, an _editor_ role might carry `(experiment, *, EDIT)` to
let its members edit every experiment, while a narrower role could combine
`(experiment, 42, READ)` with `(prompt, 7, EDIT)`.

### Permission levels

The grantable permission levels for resource-scoped permissions are:

| Permission | Can read | Can use | Can update | Can delete | Can manage |
| ---------- | -------- | ------- | ---------- | ---------- | ---------- |
| `READ`     | Yes      | No      | No         | No         | No         |
| `USE`      | Yes      | Yes     | No         | No         | No         |
| `EDIT`     | Yes      | Yes     | Yes        | No         | No         |
| `MANAGE`   | Yes      | Yes     | Yes        | Yes        | Yes        |

`USE` is for consuming a resource without modifying it (invoking a gateway
endpoint, referencing a model definition, creating new experiments /
registered models within a workspace).

`NO_PERMISSIONS` exists as the deny-by-default sentinel. It's the effective

> state of any `(user, resource)` pair with no matching grant in the resolver's

Reviewer 2: What does this mean? The default permission is configurable, and we use READ by default.

Author: You're right — "deny-by-default sentinel" was misleading. The actual default fallback is `default_permission` (`basic_auth.ini:2` ships it as `READ`), so a `(user, resource)` pair with no matching grant resolves to READ, not NO_PERMISSIONS. `a73bc583a` rewrites the paragraph: NO_PERMISSIONS is now framed as the resolver's deny sentinel for workspace-boundary cases (no membership, lookup error) — never the default. The Workspace Member third bullet also picked up the same wrong mental model; rewrote it to credit the `default_permission` floor for read access. 🤖 *Claude response*

> chain. It is **not grantable** as a role permission or a direct grant; the

Reviewer 2: q: is this correct? does the API reject when users try to assign NO_PERMISSIONS to roles?

Author: Yes — `_validate_permission_for_resource_type` (`mlflow/server/auth/permissions.py:152-179`) rejects `NO_PERMISSIONS` for both grant flavors. The resource-scope set (`RESOURCE_GRANTABLE_PERMISSIONS`, line 132) is `{READ, USE, EDIT, MANAGE}` — no NO_PERMISSIONS. The workspace-scope set (`WORKSPACE_GRANTABLE_PERMISSIONS`, line 126) is `{USE, MANAGE}` — also no NO_PERMISSIONS. The validator is called from 12 sites in `sqlalchemy_store.py` (every grant write path), so any `grant_user_permission`, `add_role_permission`, or legacy back-compat write that tries NO_PERMISSIONS raises `INVALID_PARAMETER_VALUE`. Resolved by `a73bc583a`. 🤖 *Claude response*

auth server rejects attempts to assign it.

Because grants are folded via a max (see
[Permission resolution](#permission-resolution)), RBAC has no
explicit-deny override. To restrict access to a
specific resource, grant access more narrowly (per-resource or
resource-type wildcard) rather than holding a broad workspace-level grant
and trying to except the resource.

#### Permission resolution

For each authorization check, MLflow evaluates the user's effective
permission as follows:

1. **Platform Admin bypass**: `is_admin = true` short-circuits to allow.
2. **Role-derived grants**: all of the user's roles in the request's workspace
   contribute. Grants that apply to the resource fold via a max, and the
   highest-priority permission wins (`MANAGE` > `EDIT` > `USE` > `READ`).
   `(workspace, *, MANAGE)` folds into every per-resource check — it grants
   management of every resource in the workspace. `(workspace, *, USE)`
   does **not** fold into per-resource checks; it confers workspace
   membership and create rights only (see
   [Workspace isolation](#workspace-isolation)).
3. **Server `default_permission`**: a floor max'd into the resolution

>    result. `NO_PERMISSIONS` (explicit deny) is preserved — the floor does

Reviewer 2: Iirc, this statement conflicts with the explanation above

Author: Good catch — calling NO_PERMISSIONS "explicit deny" in step 3 conflicted with the paragraph above saying it's not grantable. `a73bc583a` drops the "explicit deny" framing: step 3 now says the floor doesn't lift NO_PERMISSIONS "when the resolver returns it for a workspace-boundary reason (no workspace membership, missing workspace, lookup error)" — which matches where NO_PERMISSIONS actually originates (`mlflow/server/auth/__init__.py` lines 458 / 470 / 479). Step 3 also now says `default_permission` defaults to `READ` so the floor's typical value is concrete. 🤖 *Claude response*

  - Does **not** grant per-resource read access. Resource-level permissions
    still require an explicit grant (per-resource, resource-type wildcard,
    or the server `default_permission` floor).
- `MANAGE` on `(workspace, *)` — the **Workspace Admin** grant. Full
  authority within the workspace, including creating roles, granting
  permissions, and managing role assignments. Cannot perform system-wide
  operations such as deleting users.

The seeded `admin` and `user` roles each carry one of these grants; see
[Default roles](#default-roles).

### User tiers

The admin UI uses **Platform Admin** for the first tier and **Workspace
Admin** for the second; the doc uses the same labels.

| Tier            | How it's expressed                          | Capability                                                                     |
| --------------- | ------------------------------------------- | ------------------------------------------------------------------------------ |
| Platform Admin  | `is_admin = true` on the user row           | Unrestricted system-wide. Sole bearer of user delete and bulk operations.      |

> | Workspace Admin | Holds `(workspace, *, MANAGE)` via any role | Full authority within those workspaces; can manage roles, users, and grants.   |

Reviewer 2: Workspace manager?

Author: You're right — the UI says **Workspace Manager**, not Workspace Admin. Verified across `mlflow/server/js/src/admin/route-defs.ts:20`, `AdminPage.tsx:442/519/543`, `UserDetailPage.tsx:202`, `AccountPage.tsx:417`, and three i18n entries in `en.json`. `a73bc583a` renames "Workspace Admin" → "Workspace Manager" throughout the RBAC doc (7 occurrences). The other two docs (`permissions.mdx`, `basic-http-auth.mdx`) already had it right. (Round 3 had me rename in the other direction based on a Copilot misread of the UI; reverting now.) 🤖 *Claude response*

<TabItem value="ui" label="Admin UI — from the role">

1. Navigate to `/admin` → **Roles** → click the role.
2. Click **Edit role**.
3. In _Assigned users_, add all the users.
4. Review → apply.

</TabItem>
<TabItem value="ui-per-user" label="Admin UI — from each user">

1. Navigate to `/admin` → **Users** → click the user.
2. Click **Edit access**.
3. In _Role assignments_, add the role → apply.
4. Repeat for each user.

One round trip per user end-to-end. Pick whichever fits the operator's
mental model; the result is identical.

</TabItem>
<TabItem value="api" label="Role API">

Loop over the team and call `assign_role` once per member:

```python
user_role = next(r for r in auth_client.list_roles("ml-research") if r.name == "user")
for member in ("alice", "bob", "carol"):
    auth_client.assign_role(username=member, role_id=user_role.id)
```

</TabItem>
</Tabs>

## Admin UI

The admin UI is the operator-facing surface for everything above. It's
reached from two entry points:

- **Platform Admins** (`is_admin = true`) navigate to `/admin` via the
  sidebar `Manage` entry. The page renders cross-workspace: every user in
  the system, every role in every workspace.

> - **Workspace Admins** click a gear icon on the home-page workspaces table

Reviewer 2: ditto

Author: Same as above — covered by the rename in `a73bc583a`. Admin UI section now says "Workspace Managers click a gear icon …". 🤖 *Claude response*
